### PR TITLE
feat: [PL-39278]: Add scope filter param in activity history api

### PR DESCRIPTION
**Pull request by tiwaritejas**
## Describe your changes
This PR adds a query param- "scopeFilter" to the /activityHistory API. This is needed to return the activities where the referredByEntity belongs to the specified scopes. 
Testing-
1. Tested that passing the param as null returns all activities.
2. Tested that when the param is specified, only the activities belonging to those scopes are returned.
3. Added unit testing.
## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [x] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/51217)

### CODE REVIEW

**Comment by tiwaritejas** [comment]
trigger uts

**Review by nishantsaini20 — CHANGES_REQUESTED** [review]
> Reviewed 10 of 10 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/51217)* status: 10 file(s) to review. 1 change requests. 0 of 2 approvals obtained. 

___
*[`950-ng-core/src/main/java/io/harness/ng/core/activityhistory/entity/NGActivity.java` line 69 at r1](https://reviewable.io/reviews/harness/harness-core/51217#-NavFOjK8URvMYO27pMt:-NavFOjK8URvMYO27pMu:bntx303) ([raw file](https://github.com/harness/harness-core/blob/7f11cabfbe72b26a2b293b07cea0c093aa4289e6/950-ng-core/src/main/java/io/harness/ng/core/activityhistory/entity/NGActivity.java#L69)):*
> ```Java
>   public static List<MongoIndex> mongoIndexes() {
>     return ImmutableList.<MongoIndex>builder()
>         .add(CompoundMongoIndex.builder()
> ```

Check if index is required to support scope filter.

**Review by tiwaritejas — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/51217)* status: Some of the checks are too old. 10 file(s) to review. 1 change requests. 1. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`950-ng-core/src/main/java/io/harness/ng/core/activityhistory/entity/NGActivity.java` line 69 at r1](https://reviewable.io/reviews/harness/harness-core/51217#-NavFOjK8URvMYO27pMt:-NbOl---AsXiMjApQjzm:bv0eqfv) ([raw file](https://github.com/harness/harness-core/blob/7f11cabfbe72b26a2b293b07cea0c093aa4289e6/950-ng-core/src/main/java/io/harness/ng/core/activityhistory/entity/NGActivity.java#L69)):*
<details><summary><i>Previously, nishantsaini20 (Nishant Saini) wrote…</i></summary><blockquote>

Check if index is required to support scope filter.
</blockquote></details>

Adding index for referredByEntity scope does not help further here, as it is applies a simple filter on the scope fields. This might be because of null check in the query, not sure though

**Review by tiwaritejas — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/51217)* status: 17 file(s) to review. 1 change requests. 1. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`950-ng-core/src/main/java/io/harness/ng/core/activityhistory/entity/NGActivity.java` line 69 at r1](https://reviewable.io/reviews/harness/harness-core/51217#-NavFOjK8URvMYO27pMt:-NeNUy6B45rR0DUO6nTg:b-pf1nfc) ([raw file](https://github.com/harness/harness-core/blob/7f11cabfbe72b26a2b293b07cea0c093aa4289e6/950-ng-core/src/main/java/io/harness/ng/core/activityhistory/entity/NGActivity.java#L69)):*
<details><summary><i>Previously, tiwaritejas (Tejas Tiwari) wrote…</i></summary><blockquote>

Adding index for referredByEntity scope does not help further here, as it is applies a simple filter on the scope fields. This might be because of null check in the query, not sure though
</blockquote></details>

Added a scope field as well added index for that.

File: 970-ng-commons/src/main/java/io/harness/beans/EntityReference.java [hunk 1]
```diff
@@ -20,6 +20,7 @@
     subTypes = {IdentifierRef.class, InputSetReference.class, NGTemplateReference.class, TriggerReference.class},
     discriminator = "type")
 public interface EntityReference extends NGAccess {
+  io.harness.encryption.Scope getScope();
```
Comment (yogesh-chauhan) [inline 970-ng-commons/src/main/java/io/harness/beans/EntityReference.java:23]: nit: do we need FQNed import ?

File: 970-ng-commons/src/main/java/io/harness/beans/InfraDefReference.java [hunk 2]
```diff
@@ -39,4 +43,12 @@ public String getFullyQualifiedName() {
   public Boolean isDefault() {
     return isDefault;
   }
+
+  @Override
+  public io.harness.encryption.Scope getScope() {
+    if (!isNull(this.scope)) {
+      return this.scope;
+    }
+    return io.harness.encryption.Scope.of(accountIdentifier, orgIdentifier, projectIdentifier);
```
Comment (yogesh-chauhan) [inline 970-ng-commons/src/main/java/io/harness/beans/InfraDefReference.java:52]: nit: 
```java
return scope == null ? Scope.of(..) : scope 
```

Also, FQN import is not needed

**Review by yogesh-chauhan — DISMISSED** [review]

**Review by nishantsaini20 — DISMISSED** [review]
> Reviewed 1 of 10 files at r1, 17 of 17 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/51217)* status: Some of the checks are too old. 17 file(s) to review.

**Comment by tiwaritejas** [comment]
trigger uts

**Comment by tiwaritejas** [comment]
trigger uts

**Comment by tiwaritejas** [comment]
trigger uts

**Review by nishantsaini20 — COMMENTED** [review]
> Reviewed 6 of 6 files at r3, 2 of 2 files at r4, 1 of 1 files at r5.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/51217)* status: 19 file(s) to review. 0 of 2 approvals obtained.

**Review by nishantsaini20 — DISMISSED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/51217)* status: 19 file(s) to review. 1 of 2 approvals obtained.

**Comment by yogesh-chauhan** [comment]
*[`950-ng-core/src/test/java/io/harness/ng/core/activityhistory/ActivityHistoryTestHelper.java` line 37 at r5](https://reviewable.io/reviews/harness/harness-core/51217#-Nf1dTrbBSDayDIkpdrS:-Nf1dTrbBSDayDIkpdrT:b2fp7ht) ([raw file](https://github.com/harness/harness-core/blob/0b287200b2c22858990c9f966cab3364d2815b7c/950-ng-core/src/test/java/io/harness/ng/core/activityhistory/ActivityHistoryTestHelper.java#L37)):*
> ```Java
> 
>     if (isEmpty(referredByEntityIdentifier)) {
>       if (referredByEntityScope.equals(Scope.PROJECT)) {
> ```

nit: its a good practice to follow to put constants on left side of comparison
https://mkyong.com/java-best-practices/constant-value-should-always-come-first-in-comparison/

**Review by yogesh-chauhan — COMMENTED** [review]
> Reviewed 7 of 17 files at r2, 5 of 6 files at r3, 2 of 2 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/51217)* status: 4 file(s) to review. 1 of 2 approvals obtained.

**Comment by tiwaritejas** [comment]
trigger uts

**Comment by tiwaritejas** [comment]
trigger uts

**Comment by tiwaritejas** [comment]
trigger uts

**Review by yogesh-chauhan — APPROVED** [review]
> Reviewed 2 of 17 files at r2, 4 of 4 files at r6, 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/51217)* status: 5 file(s) to review. 1 of 2 approvals obtained.

**Review by archit-harness — APPROVED** [review]
> Reviewed 6 of 17 files at r2, 5 of 6 files at r3, 2 of 2 files at r4, 1 of 1 files at r5, 4 of 4 files at r6, 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/51217)* status: :shipit: complete! ✓

**Pull request merged**